### PR TITLE
Start encapsulating submaps and simplify their rotation

### PR DESCRIPTION
**Pull request by codemime**
#### Summary

```SUMMARY: Infrastructure "Starts encapsulating submaps and simplifies their rotation"```

#### Purpose of change

The ultimate goal is #29674, but robust implementation requires proper encapsulation of `submap`. This PR starts it.

#### Describe the solution

- Fixed a regression in map editor (0809ffa), introduced in #29161. It stayed broken, but unreported ever since. I wonder if it's used at all anyways?
- Simplified submap placement in map editor (05e513d) to avoid manipulating raw array pointers which prevented encapsulation;
- Simplified submap rotation, moved it from `map` to `submap`, and added very basic unit tests (8ce55f8);
- Relocated serialization from `overmap_buffer` to `submap`. The implementation is conventionally put into `savegame_json.cpp` (fbf2fae);
- Turned `submap` into a class (057db6b): we want private members there;
- Encapsulated submap temperature (6db69ee).

#### Describe alternatives you've considered
None.

#### Additional Context
No intended changes in observable behavior.

### CODE REVIEW

File: src/editmap.cpp [hunk 1]
```diff
@@ -1973,7 +1931,7 @@ int editmap::edit_mapgen()
         gmenu.addentry( -1, !id.id().is_null(), 0, "[%3d] %s", static_cast<int>( id ), id.id().c_str() );
         gmenu.entries[i].extratxt.left = 1;
         gmenu.entries[i].extratxt.color = id->get_color();
-        gmenu.entries[i].extratxt.txt = string_format( "%c", id->get_symbol() );
+        gmenu.entries[i].extratxt.txt = id->get_symbol();
```
Comment (ZhilkinSerg) [inline src/editmap.cpp]: Good catch - that's was definitely an error. I think this is from RARELY used debug menu and that's is why it wasn't caught earlier.

Broken:

image


Fixed:

image

**Comment by ZhilkinSerg** [comment]
Attached savegame triggers crash. Reproduce steps:

1. Debug reveal overmap
2. Debug teleport to center of nearest city north of player location
3. Debug kill all monsters
4. Try to move player north.

[Bellevue.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/3136815/Bellevue.zip)

**Comment by codemime** [comment]
> Attached savegame triggers crash. Reproduce steps:

It turns out that I forgot about active items in the cache. Thanks for detecting the problem. Should be good now.

**Comment by codemime** [comment]
Jenkins rebuild.

**Comment by codemime** [comment]
Jenkins fails for some reason: `ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job`. I'd appreciate some help. Pinging @narc0tiq.

**Comment by narc0tiq** [comment]
I may have fixed the Jenkins PR builder issue; let's just have **Jenkins rebuild** and see if it worked.

**Comment by narc0tiq** [comment]
Ok, my guess was correct: I'd set the filter for remote refs too strict, and it was ignoring all PRs whose number didn't start with 3. I've just changed it to include "29" as a valid prefix.

Normally I wouldn't filter at all (i.e., allow all PRs), but when you `git fetch +refs/pull/*:refs/remotes/origin/pr/*` on CDDA, you end up creating an awful lot of refs, which is _so damn slow_. I may still switch to that, just to avoid having to fix it again in the future.

**Comment by codemime** [comment]
Thank you so much, Octav. This totally worked.

> I may still switch to that, just to avoid having to fix it again in the future.

Alternatively, you can acquire a list of all pending PRs in json via github api: [https://api.github.com/repos/CleverRaven/Cataclysm-DDA/pulls?state=open](https://api.github.com/repos/CleverRaven/Cataclysm-DDA/pulls?state=open).

**Comment by narc0tiq** [comment]
Sure, but I can't inject that into Jenkins' git fetch step. It's ok, I'll
let it take its time.

**Comment by Hirmuolio** [comment]
The rotting calculations were moved out of weather.cpp in #29818 so the `get_rot_since` should be re-removed.

**Comment by codemime** [comment]
> The rotting calculations were moved out of weather.cpp in #29818 so the get_rot_since should be re-removed.

It's removed in a following (not published yet) PR. Thanks for pointing this out.

**Pull request merged**